projects/storefrontlib/cms-components/product/carousel/product-carousel-item/product-carousel-item.component.spec.ts

```diff
@@ -0,0 +1,278 @@
+    component = fixture.componentInstance;
+    componentInjector = fixture.debugElement.injector;
+
+    component.item = mockProduct;
+
+    component.ngOnChanges({});
+    fixture.detectChanges();
+  });
+
+  it('should create', () => {
+    expect(component).toBeTruthy();
+  });
+
+  it('should display product name', () => {
+    expect(
+      fixture.debugElement.nativeElement.querySelector('.cx-product-name')
+        .textContent
+    ).toContain(component.item.name);
+  });
+
+  it('should display product formatted price', () => {
+    expect(
+      fixture.debugElement.nativeElement.querySelector('.price').textContent
+    ).toContain(component.item.price.formattedValue);
+  });
+
+  it('should display product image', () => {
+    expect(
+      fixture.debugElement.nativeElement.querySelector('cx-media')
+    ).not.toBeNull();
+  });
+
+  it('should have defined instance of list item context', () => {
+    expect(component['productListItemContextSource']).toBeDefined();
+  });
+
+  it('should provide ProductListItemContextSource', () => {
+    expect(componentInjector.get(ProductListItemContextSource)).toBeTruthy();
+  });
+
+  it('should provide ProductListItemContext', () => {
+    expect(componentInjector.get(ProductListItemContext)).toBe(
+      componentInjector.get(ProductListItemContextSource)
+    );
+  });
+
+  it('should push changes of input"product" to context', () => {
+    const contextSource: ProductListItemContextSource = componentInjector.get(
+      ProductListItemContextSource
+    );
+    spyOn(contextSource.product$, 'next');
+    component.item = mockProduct;
+    component.ngOnChanges({
+      item: { currentValue: component.item } as SimpleChange,
+    });
+    expect(contextSource.product$.next).toHaveBeenCalledWith(mockProduct);
+  });
+
+  describe('UI test', () => {
+    it(
+      'should render product name in template',
+      waitForAsync(() => {
+        const el = fixture.debugElement.query(By.css('h3'));
+        expect(el.nativeElement).toBeTruthy();
+        expect(el.nativeElement.innerText).toEqual('Test product');
+      })
+    );
+
+    it(
+      'should render product price in template',
+      waitForAsync(() => {
+        const el = fixture.debugElement.query(By.css('.price'));
+        expect(el.nativeElement).toBeTruthy();
+        expect(el.nativeElement.innerText).toEqual('$100,00');
+      })
+    );
+
+    it(
+      'should render product primary image for the first item',
+      waitForAsync(() => {
+        const el = fixture.debugElement.query(By.css('cx-media'));
+        expect(el.nativeElement).toBeTruthy();
+      })
+    );
+
+    it(
+      'should render missing product image for the 2nd item as well',
+      waitForAsync(() => {
+        const el = fixture.debugElement.query(By.css('cx-media'));
+        expect(el.nativeElement).toBeTruthy();
+      })
+    );
+  });
+});
+
+describe('ProductCarouselItemComponent in product-carousel with add-to-cart component', () => {
```

Comment:
This part I think is more of an integration test (covered by the e2e) and doesn't belong in the unit tests.
As far as  ProductCarouselItemComponent is concerned, the only thing it's responsible for is to have  cxInnerComponentsHost in its template  and that on change, it emits a new item value in the context.

Reply:
I shall delete this test then - the above case is covered in lines 148 - 158 already.